File: src/main/kotlin/com/bookislife/jgsk/kotlin/_15_class/Man.kt
package com.bookislife.jgsk.kotlin._15_class

/**
 * Created by SidneyXu on 2015/06/05.
 */
//  Primary Constructor
//  Var for mutable and val for read-only
class Man(val name: String,
          var age: Int,
          private var from: String = "USA") {

    init {
        println("Sentences in primary constructor")
    }

    //  Secondary Constructor
    //  Delegate to the primary constructor
    constructor(name: String) : this(name, 0) {
    }

    constructor(age: Int) : this("Default Name", age) {
    }

    constructor() : this("Default Name") {
    }
}

class Woman private constructor(val name: String, val age: Int) {
    constructor(name: String) : this(name, 0) {

    }
}

class Empty

File: src/main/scala/com/bookislife/jgsk/scala/_15_class/Man.scala
package com.bookislife.jgsk.scala._15_class

/**
  * Created by SidneyXu on 2015/2/12.
  */
//  Main Constructor
//  Parameters without var or val in main constructor are same as private[this].
class Man(val name: String,
          var age: Int,
          private var from: String = "USA",
          description: String = "none") {

    println("Sentences in Main Constructor")

    //  Slave Constructor
    def this(name: String) {
        //  Delegate to the main constructor
        this(name, 0)
    }
  
    def this(age: Int) {
        this("Default Name", age)
    }

    def this() {
        this("Default Name")
    }

    override def toString = s"Man($name, $age, $from, $description)"
}

class Woman private(val name: String, val age: Int) {
    def this(name: String) {
        this(name, 0)
    }
}